tests/components/ihc/test_config_flow.py

```diff
@@ -0,0 +1,190 @@
+    """Test we handle invalid auth."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.ihc.config_flow.IHCController.is_ihc_controller",
+        return_value=True,
+    ), patch(
+        "homeassistant.components.ihc.config_flow.IHCController.authenticate",
+        return_value=False,
+    ), patch(
+        "homeassistant.components.ihc.config_flow.IHCController.disconnect",
+        return_value=None,
+    ), patch(
+        "homeassistant.components.ihc.async_setup_entry",
+        return_value=True,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "url": "http://1.1.1.1",
+                "username": "test-username",
+                "password": "test-password",
+                "auto_setup": True,
+            },
+        )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "invalid_auth"}
+
+
+async def test_form_cannot_connect(hass: HomeAssistant) -> None:
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.ihc.config_flow.IHCController.is_ihc_controller",
+        return_value=False,
+    ), patch(
+        "homeassistant.components.ihc.async_setup_entry",
+        return_value=True,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "url": "http://1.1.1.1",
+                "username": "test-username",
+                "password": "test-password",
+                "auto_setup": True,
+            },
+        )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "cannot_connect"}
+
+
+async def test_already_setup(hass: HomeAssistant) -> None:
+    """Test if controller is already setup."""
+    MockConfigEntry(domain=DOMAIN, unique_id="123", data={}).add_to_hass(hass)
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == "form"
+    assert result["errors"] == {}
+
+    with patch(
+        "homeassistant.components.ihc.config_flow.IHCController.is_ihc_controller",
+        return_value=True,
+    ), patch(
+        "homeassistant.components.ihc.config_flow.IHCController.authenticate",
+        return_value=True,
+    ), patch(
+        "homeassistant.components.ihc.config_flow.IHCController.disconnect",
+        return_value=None,
+    ), patch(
+        "homeassistant.components.ihc.config_flow.get_controller_serial",
+        return_value="123",
+    ), patch(
+        "homeassistant.components.ihc.async_setup_entry",
+        return_value=True,
+    ):
```

Comment:
```suggestion
    ):
```

Reply:
Done, moved async_setup_entry  mock to fixture